File: Simulated Sensor Data Logger.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>  // For sleep()

// Simulate sensor readings
float getTemperature() {
    return 20.0 + (rand() % 1000) / 100.0;  // 20.00 to 40.00 °C
}

float getHumidity() {
    return 30.0 + (rand() % 500) / 10.0;    // 30.0 to 90.0 %
}

int getLightIntensity() {
    return rand() % 1001;                  // 0 to 1000 Lux
}

void logData(FILE *fp, float temp, float humidity, int light) {
    time_t now;
    time(&now);
    char *timestamp = ctime(&now);
    timestamp[strcspn(timestamp, "\n")] = 0;  // remove newline

    fprintf(fp, "%s - Temp: %.2f °C, Humidity: %.1f %%, Light: %d Lux\n",
            timestamp, temp, humidity, light);
    fflush(fp);  // Ensure it's written immediately
}

int main() {
    FILE *fp = fopen("sensor_log.txt", "a");
    if (!fp) {
        printf("Error opening file!\n");
        return 1;
    }

    srand(time(NULL)); // Seed random number generator

    printf("Simulating sensor data logging...\n");

    for (int i = 0; i < 10; i++) { // Simulate 10 readings
        float temp = getTemperature();
        float humidity = getHumidity();
        int light = getLightIntensity();

        logData(fp, temp, humidity, light);

        sleep(1); // Wait for 1 second before next reading
    }

    fclose(fp);
    printf("Sensor data logging complete. Check 'sensor_log.txt'.\n");
    return 0;
}

File: Simulated Sensor Data Logger.py
import random
import time
from datetime import datetime

# Simulate temperature sensor (20°C to 30°C)
def get_temperature():
    return round(random.uniform(20.0, 40.0), 2)

# Simulate humidity sensor (30% to 80%)
def get_humidity():
    return round(random.uniform(30.0, 90.0), 1)

# Simulate light sensor (0 to 1000 Lux)
def get_light_intensity():
    return random.randint(0, 1000)

# Log data to file with timestamp
def log_data(temp, humidity, light):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    log_line = f"{timestamp} - Temp: {temp}°C, Humidity: {humidity}%, Light: {light} Lux"
    
    # Print to console
    print(log_line)

    # Append to file
    with open("sensor_log.txt", "a") as file:
        file.write(log_line + "\n")

# Main function
def main():
    print("Simulated Sensor Data Logger Running...\n")

    for _ in range(10):  # Simulate 10 readings
        temp = get_temperature()
        humidity = get_humidity()
        light = get_light_intensity()

        log_data(temp, humidity, light)
        time.sleep(1)  # 1-second interval

    print("\nLogging complete. Check 'sensor_log.txt'.")

if __name__ == "__main__":
    main()
